lib/Basics/StringUtils.h

```diff
@@ -418,6 +418,101 @@ std::string decodeHex(std::string const& value);
 void escapeRegexParams(std::string& out, const char* ptr, size_t length);
 std::string escapeRegexParams(std::string const& in);
 
+namespace impl {
+template <typename T>
+auto constexpr isStringOrView = std::is_same_v<std::string, std::decay_t<T>> ||
+                                std::is_same_v<std::string_view, std::decay_t<T>>;
+
+template <typename T>
+auto toStringOrView(T&& arg) {
+  using Arg = std::decay_t<T>;
+  if constexpr (std::is_same_v<std::string, Arg> || std::is_same_v<std::string_view, Arg>) {
+    return arg;
+  } else if constexpr (std::is_convertible_v<Arg, std::string_view>) {
+    return std::string_view(arg);
+  } else if constexpr (std::is_convertible_v<Arg, std::string>) {
+    return std::string(arg);
+  } else {
+    // Use using, so ADL could also find to_string in other namespaces than std.
+    using std::to_string;
+    return to_string(arg);
+  }
+}
+
+template <typename... Iters>
+auto concatImplIter(std::pair<Iters, Iters>&&... iters) -> std::string {
+  auto result = std::string{};
+
+  result.reserve((std::distance(iters.first, iters.second) + ... + 0));
+
+  ([&] { result.append(iters.first, iters.second); }(), ...);
+
+  return result;
+}
+
+/// @brief Converts all arguments to a pair of iterators (begin, end), passing
+/// them to concatImplIter.
+/// All arguments must either be `std::string` or `std::string_view`.
+template <typename... Args>
+auto concatImplStr(Args&&... args) -> std::string {
+  static_assert(((isStringOrView<Args>)&&...));
+  return concatImplIter(std::make_pair(args.begin(), args.end())...);
+}
+
+template <typename Iter, typename... Iters>
+auto joinImplIter(std::string_view delim, std::pair<Iter, Iter>&& head,
+                  std::pair<Iters, Iters>&&... tail) -> std::string {
+  auto result = std::string{};
+
+  auto const valueSizes = std::distance(head.first, head.second) +
+                          (std::distance(tail.first, tail.second) + ... + 0);
+  auto const delimSizes = sizeof...(Iters) * delim.size();
```

Comment:
shouldn't there be `sizeof...(Iters) - 1` many delims?

Reply:
No, there are `1 + sizeof...(Iters)` many elements: note `head` and `tail`.

Reply:
👍🏽